src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpResponseHeaderReader.cs

```diff
@@ -0,0 +1,168 @@
+                {
+                    name = new string(_buffer, startIndex, nameLength);
+                }
+
+                // Normalize header value by trimming white space.
+                int valueStartIndex = colonIndex + 1;
+                int valueLength = startIndex + length - colonIndex - 1;
+                CharArrayHelpers.Trim(_buffer, ref valueStartIndex, ref valueLength);
+
+                if (valueLength == 0)
+                {
+                    value = string.Empty;
+                }
+                else
+                {
+                    // If it's a known header value, use the known value instead of allocating a new string.
+                    if (!TryGetKnownHeaderValue(name, _buffer, valueStartIndex, valueLength, out value))
+                    {
+                        value = new string(_buffer, valueStartIndex, valueLength);
+                    }
+                }
+
+                return true;
+            }
+
+            name = null;
+            value = null;
+            return false;
+        }
+
+        /// <summary>
+        /// Reads lines separated by "\r\n".
+        /// </summary>
+        /// <returns>true if the next line was read successfully, or false if all characters have been read.</returns>
+        public bool ReadLine()
+        {
+            int startIndex;
+            int length;
+            return ReadLine(out startIndex, out length);
+        }
+
+        /// <summary>
+        /// Reads lines separated by "\r\n".
+        /// </summary>
+        /// <param name="startIndex">The start of the line segment.</param>
+        /// <param name="length">The length of the line segment.</param>
+        /// <returns>true if the next line was read successfully, or false if all characters have been read.</returns>
+        private bool ReadLine(out int startIndex, out int length)
+        {
+            Debug.Assert(_buffer != null);
+
+            int i = _position;
+
+            while (i < _length)
+            {
+                char ch = _buffer[i];
+                if (ch == '\r')
+                {
+                    int next = i + 1;
+                    if (next < _length && _buffer[next] == '\n')
+                    {
+                        startIndex = _position;
+                        length = i - _position;
+                        _position = i + 2;
+                        return true;
+                    }
+                }
+                i++;
+            }
+
+            if (i > _position)
+            {
+                startIndex = _position;
+                length = i - _position;
+                _position = i;
+                return true;
+            }
+
+            startIndex = 0;
+            length = 0;
+            return false;
+        }
+
+        private static bool TryGetKnownHeaderValue(string name, char[] array, int startIndex, int length, out string value)
```

Comment:
It could certainly be fleshed out more if there are other headers with common known values. I didn't spend much time investigating whether there are any others like that.